Changes in the mei-Mensural.xml customization to make it compliant with the decisions made by the mensural-ig

martha-thomae opened the pull request:
Three main changes:

1. New attribute `@dur.quality` for `<note>` elements to indicate whether the note is perfect / imperfect / altered / etc. and, therefore, be comprehensive in the durational information of the note (given that in mensural notation, the note shape is not enough to indicate the length of a note). The new attribute can have the following six values: `perfecta`, `imperfecta`, `altera`, `maior`, `minor`, and `duplex`. The last three are used in Ars antiqua: `maior` and `minor` are used for semibreves and `duplex` is used for longae.

2. New values for the attribute `@dur` for the `<rest>` elements: two-breve rests ("2B") and three-breve rests ("3B"). 

3. New `<stem>` element to be used as children of the `<note>` element. Notes can have more than one stem. Therefore the introduction of the `<stem>` element as a children of `<note>`, allowing to associate multiple stem elements to a single note, each stem with its own form, direction, etc. Attributes of `<stem>`: `pos`, `length`, `form`, `dir`, `flag.pos`, `flag.form`, `x`, and `y`.


Two small changes:

4. Change in the description of the `@dur` attribute

5. Added a new stem-related attribute `@stem.form` to be encoded within the `<note>` element.

These are all (but one) of the changes that were discussed and approved by the mensural-ig in the 2018 meeting. The one change I did not include has to do with the encoding of the mensuration sign, because (based on the notes) further discussion is needed. 
**[The discussion can be seen in the attached file.]**
[MEI-Mensural IG - Approved MEC 2018 Pending Implementation.pdf](https://github.com/music-encoding/music-encoding/files/3853621/MEI-Mensural.IG.-.Approved.MEC.2018.Pending.Implementation.pdf)

martha-thomae committed:
Adding the dur.quality attribute for the mensural module

martha-thomae committed:
Change in the descriptions related to attributes dur and dur.quality.
Regarding the attribute dur: change in the general description (dur refers to the note symbol, not to the duration of the note) and change in the description of the 'brevis', 'semiminima', 'fusa', and 'semifusa'.
Regarding the attribute dur.quality: Added a general description of the data type and attribute, and modified the description of the individual values to be more accurate.

martha-thomae committed:
Adding missing durational values for rests.
Allowing the encoding of two-breve and three-breve rests (this values are only allowed for rests, not notes).

martha-thomae committed:
Adding stem element and more stem-related attributes.
Given that mensural notes can have two stems (with different directions and their own forms and flags), the mensural-ig approved the idea of introducing the 'stem' element as a child of the 'note' element. The changes here allow for the use of this 'stem' element and for its attributes: pos, length, form, dir, flag.pos, flag.form, x, and y.
The mensural-ig also approved the introduction of the 'stem.form' attribute to be used within the note element (the other attributes corresponding to position, length, direction, and x and y coordinates are already included in the current specification of MEI).

kepper reviewed (commented):
Thanks for this proposal. I have some questions and issues left. First, personalized comments should not be included. Second, if comments contain useful information, that should go into `<desc>` or `<remarks>` instead – we should try to avoid comments as much as possible, as they can't be processed (easily). Then, ultimately, new attributes shouldn't be added to a customization, but to the sources. This assumes that there is consensus on what should be added, so testing things inside a customization is perfectly appropriate. Regarding the proposal itself: I wonder if the name `@dur.quality` could be misleading for people not dealing with the mensural stuff. First, I stepped into this question, and was bringing arguments against that. However, when people turn off the mensural module, they won't see it, so when using MEI properly, this shouldn't be an issue. I would still like to hear others' opinions on this… Schedule-wise: the Tech Chairs agreed that we should come up with regular (virtual) meetings where PRs can be discussed. Only when there is consensus at such a meeting, things should get pulled. I will send out an email about that procedure today, so I would ask to use the time until then (Nov 29) to address the questions above. Thanks!

martha-thomae commented:
Hi @kepper. So:

1. I will remove all comments (erase the personalized ones and move the more informative ones into either `<desc>` or `<remark>`).
2. I know there have been issues regarding the naming of the attribute as `@dur.quality`, but in the end (after several discussions in the mensural-ig) no other name has been proven more appropriate. And yes, the idea is that it would only be included within the mensural module. And also the documentation shown when you are adding the attribute (in Oxygen) shows that this is an attribute for mensural notation.
3. So, just to be clear, should I include these changes in the customization file or the source file at this point? I will redo the pull-request to remove the comments, so I can make the changes in the appropriate file as well.

Please let me know if there is anything else needed before the 29th. I will try to have everything ready so that this PR can be approved. Thanks!

martha-thomae committed:
Remove comments from the mei-Mensural.xml

giulianodibacco commented:
About the individual points: 
#1 - the IG has debated for months about the best and concise attribute name, and "quality" has been approved as the most appropriate. There is no Latin original for this. "Quality" in English is "a distinctive attribute or characteristic possessed by someone or something" so in this case we want to distinguish a specific characteristic of "duration". Note that if we wanted to split hairs, what we encode with "dur" in mensural terms is not really a "duration" in the same way it is for CMN; but we don't want to use a completely different system to represent how time is "measured", right? I think that mensuralists should compromise in order to keep a core set of tags common to all modules, and this is what is happening here. "Quality" is an excellent compromise and a term generic enough that everybody will understand. The Guidelines will explain.
#2 - I am not a fan of "2B" and "3B" but these values work for what we need, and people seem to like these values. Also in this case the discussion led to agree about this so I hope that this can be introduced.
#3, 5 - As the description says (and as the original document of the IG further explains) notes can have more than one stem, with associated flags of various shapes. I cannot think to a simplest way than an element to represent this.

The original document approved by the IG suggests that in order not to bother people not encoding mensural stuff, special the use of elements or attributes that can generate confusion can be restricted via schematron. How useful to have notationtype.

I have a question about the procedure. Wouldn't it be more practical to have one pull request per each feature, so that each of them can be separately discussed, if needed? This PR deal with too different matters to be discussed together, examples made.

kepper reviewed (approved):
Everything that I've requested has been addressed, so this should be fine to merge. What I haven't done, though, is to compile and play with the resulting scheme to see if unintentional side effects where introduced. However, we should trust the IG not only for picking the right terminology, but also that they've checked their own work…

martha-thomae commented:
As discussed in the ODD Friday, I would prefer that someone else (other than me) would actually play around with the resulting scheme. This is my first time editing the ODD and I would want to double-check that these changes don't have, as you call them, unintentional side effects. Thank you for offering to take a look into this in a few weeks, @kepper. I will proceed to include some example images to show the intended use of the introduced features (to make this task a little easier).

martha-thomae commented:
The first change indicated in the pull request is about the `@dur.quality` attribute, the description in the PR indicates:

> New attribute @dur.quality for <note> elements to indicate whether the note is perfect / imperfect / altered / etc. and, therefore, be comprehensive in the durational information of the note (given that in mensural notation, the note shape is not enough to indicate the length of a note). The new attribute can have the following six values: perfecta, imperfecta, altera, maior, minor, and duplex. The last three are used in Ars antiqua: maior and minor are used for semibreves and duplex is used for longae.


This is what I expect:

![dur quality](https://user-images.githubusercontent.com/13948831/73861037-7cddb980-480a-11ea-8f2a-2dcbbcb3b46b.png)
It is one of the available attributes of a `<note>` element

![dur quality values](https://user-images.githubusercontent.com/13948831/73861050-7f401380-480a-11ea-9a90-955a33ef0692.png)
And these are all the available values of the attribute. The guidelines, as well as `<desc>` for each of these values in the schema, will indicate when using each of these values (e.g., `major` and `minor` should only be used for `semibrevis` in _Ars antiqua_).

martha-thomae commented:
The second change indicated in the pull request reads:

> New values for the attribute `@dur` for the `<rest>` elements: two-breve rests ("2B") and three-breve rests ("3B").

This is what I expect:

The rests have these extra values (`2B` and `3B`) for the `@dur` attribute.
![dur of rests](https://user-images.githubusercontent.com/13948831/73865894-f62cda80-4811-11ea-98c8-1adfc0f3c484.png)

While these values are not available in the notes.
![dur of notes](https://user-images.githubusercontent.com/13948831/73865887-f4fbad80-4811-11ea-95a0-72c0e5cdc3f3.png)

martha-thomae commented:
The third change indicated in the pull request reads:

> New `<stem>` element to be used as children of the `<note>` element. Notes can have more than one stem. Therefore the introduction of the `<stem>` element as a child of `<note>`, allowing to associate multiple stem elements to a single note, each stem with its own form, direction, etc. Attributes of `<stem>`: `pos`, `length`, `form`, `dir`, `flag.pos`, `flag.form`, `x`, and `y`.

The values for the `@form`, `@flag.pos`, and `@flag.form` were defined from scratch in this pull request. The values of the other attributes are taken from pre-existing data types.

**dir**
![dir](https://user-images.githubusercontent.com/13948831/73867346-8d932d00-4814-11ea-9735-73ddfc6e2fa6.png)

**form**
![form](https://user-images.githubusercontent.com/13948831/73867353-91bf4a80-4814-11ea-9731-f00eaf152660.png)

**pos**
![pos](https://user-images.githubusercontent.com/13948831/73867363-9421a480-4814-11ea-9348-eaec80ac3d51.png)

**flag.form**
![flag form](https://user-images.githubusercontent.com/13948831/73867370-95eb6800-4814-11ea-9d24-fbfbfbba90c5.png)

**flag.pos**
![flag pos](https://user-images.githubusercontent.com/13948831/73867376-984dc200-4814-11ea-8942-aca416f68887.png)

**And the length, x, and y attributes use their regular values.**
![x, y, length](https://user-images.githubusercontent.com/13948831/73867380-9a178580-4814-11ea-8778-bebfdd67fbc7.png)

martha-thomae committed:
Add 'add' and 'del' mode to clarify the changes in the schema.

martha-thomae committed:
Replace handling of 2B and 3B rests with schematron rule.
Reduced the number of data types and attribute classes introduced.

martha-thomae committed:
Update .travis.yml

martha-thomae commented:
There are two other pull requests regarding the Mensural notation module: #541 and #537. About each of them in relation to this pull request:

- The #541 pull request proposes to remove mensur- and proper-related attributes from `<staffDef>` and `<scoreDef>` elements. There are no conflicts between the present PR (#634) and #541. 

- The #537 pull request introduces support for , there were some issues raised for this proposal, which I think can be resolved by moving the attributes that are related to mensuration signs into the visual domain. This was proposed by @rettinghaus in https://github.com/music-encoding/music-encoding/pull/537#issuecomment-409260333:
> So why is the sign attribute part of the logical domain? The sign itself doesn't have to carry a meaning. And the values "O" and "C" are just imitating the look of the used sign. "O" and "C" are a pretty straight forward encoding method nevertheless. **I think `@sign` and `@dots` should be in the visual domain.**

I will make this changes and submit another pull request for this (I don't want to mix it with this one in case there are some otehr issues to be rasied)
Further discussion might be needed for #537, so I will include these changes on another pull request (fixing the issues raised on #537).

giulianodibacco commented:
Just a question about the procedure. Does "changes approved" and "all checks have passed" mean that the discussion is closed, and that all the proposed changes are now part of the specification? @kepper @martha-thomae

martha-thomae commented:
Hi @giulianodibacco. Let me explain. So, @kepper accepted all the changes trusting that the IG had checked that there were no unintentional side effects introduced with these changes. I asked Johannes to please take a look in more detail to the PR because this is my first time editing ODD. So, no, the changes are not part of the schema yet. Hopefully, they will be incorporated soon.

kepper committed:
minor changes

kepper commented:
I've requested a couple of minor changes upstream over at https://github.com/martha-thomae/music-encoding/pull/1. Nothing serious…

martha-thomae committed:
Merge pull request #1 from kepper/mensural

minor changes

martha-thomae commented:
> I've requested a couple of minor changes upstream over at [martha-thomae#1](https://github.com/martha-thomae/music-encoding/pull/1). Nothing serious…

Thank you very much, @kepper, for the time you took evaluating this proposal. Thanks for the pull request you made, all the minor changes you asked in that PR have been included (previous commit).

martha-thomae committed:
Change handling of multi-breve rests.
- Substituted the schematron rule by the use of two datatype classes. One is the data type for just the '2B' and '3B' values of multi-breve rests (data.MULTIBREVERESTS.mensural), and the other is a 'meta' data type for all values of mensural rests (data.DURATIONRESTS.mensural). The latter includes both the multi-breve rests' data type ('2B' and '3B' from data.MULTIBREVERESTS.mensural) and the regular dur values ('maxima', 'longa', 'brevis', 'semibrevis', etc. contained in data.DURATION.mensural).
- Other items were introduced (data.DURATIONRESTS, att.restduration.logical) or changed (att.rest.log) to be able to use these values within the 'rest' elements without making them available in the 'note' elements as well. The data.DURATIONRESTS is substituting here the data.DURATION (that only contained the 'dur' values for notes, excluding the values of mensural multi-breve rests), and the att.restduration.logical is substituting here the att.duration.logical (that included the values given in the data.DURATION datatype). And att.rest.log is now including att.restduration.logical rather than att.duration.logical.

martha-thomae commented on customizations/mei-Mensural.xml:
```diff
@@ -203,7 +189,63 @@
+            </attDef>
+          </attList>
+        </classSpec>
+        <classSpec ident="att.rest.log" module="MEI.shared" type="atts" mode="change">
```
I don't know if the creation of these two new items **`att.restduration.logical`** and **`data.DURATIONRESTS`** was the most appropriate solution for allowing the "2B" and "3B" values to be used in `<rest>` elements without making them available in the `<note>` elements too. But here is my reasoning behind their use:

The **`att.duration.logical`** class included in **`att.rest.log`** does not include values for encoding "2B" and "3B" rests. Therefore, I substituted it (in the **`att.rest.log`**) by a newly created one: **`att.restduration.logical`**. This new class does include all rests types since it uses the data type **`data.DURATIONRESTS`** which includes both the durations of CMN (**`data.DURATION.cmn`**) and Mensural rests (**`data.DURATIONRESTS.mensural`**).

I didn't want just to add the "2B" and "3B" values to the original **`att.duration.logical`** because that would have made those values available for `<note>` elements too. 

The new **`att.restduration.logical`** and **`data.DURATIONRESTS`** (for rests) are similar in their behavior to the corresponding **`att.duration.logical`** and **`data.DURATION`** for notes. The code of the latter two can be seen here for comparison purposes:
```xml
<macroSpec ident="data.DURATION" module="MEI" type="dt">
  <desc>Logical, that is, written, duration attribute values.</desc>
  <content>
    <rng:choice>
      <rng:ref name="data.DURATION.cmn"/>
      <rng:ref name="data.DURATION.mensural"/>
    </rng:choice>
  </content>
</macroSpec>

<classSpec ident="att.duration.logical" module="MEI.shared" type="atts">
  <desc>Attributes that express duration in musical terms.</desc>
  <attList>
    <attDef ident="dur" usage="opt">
      <desc>Records the duration of a feature using the relative durational values provided by the
        data.DURATION datatype.</desc>
      <datatype>
        <rng:ref name="data.DURATION"/>
      </datatype>
    </attDef>
  </attList>
</classSpec>
```

martha-thomae commented:
Hi @kepper! Just to let you know that all the things discussed in the last ODD Thursday are now corrected here. I added a few in-line comments with questions on the code of the last commit to evaluate if the last change requested was implemented alright. Please let me know if there is anything that I have to change. 

This (#634) and pull requests #541 and #657 complete all the changes for the mensural module. There are no conflicting changes between any of these pull requests. The documentation associated with the modified specs (in these 3 pull requests) can be found in the guidelines https://github.com/music-encoding/guidelines/pull/156 PR. Everything is ready for tomorrow's ODD Friday. Please let me know if there is anything I can do. Thanks! See you tomorrow.

Oh, and when the changes on this (#634) pull request are accepted, I can move things from customization to the source files (as we agreed).

martha-thomae commented:
Everything in this PR has been double-checked and accepted, except for the last commit about the multi-breve rests. @kepper, you asked me to change their implementation: substituting the schematron rule by the use of two datatype classes (see [commit message](https://github.com/music-encoding/music-encoding/pull/634/commits/6be4736b33d3586992208322b1b66718a55d53df)). I hope I have done this correctly, I added a comment on that just after that commit.

I have already moved all the approved changes into the source files (instead of using the customization file as happens in this PR), as can be seen in [this branch](https://github.com/martha-thomae/music-encoding/tree/mensural_source). As soon as the multi-breve rest implementation is evaluated, I will also move these changes into the source files (at that new branch) and finalize the PR.
